protocol-select/README.md

```diff
@@ -0,0 +1,541 @@
+which protocol is used.
+
+#### Protocol Differentiation
+
+Both Protocol Select and [Multistream Select] prefix their messages with the
+varint encoded message length. The first message send by [Multistream Select] is
+`/multistream/1.0.0`. Implementations should read the first few bytes and
+proceed with either [Multistream Select] or Protocol Select depending on whether
+it equals `/multistream/1.0.0` or not.
+
+## Protocol Specification
+
+Messages are encoded according to the Protobuf definition below using the
+`proto2` syntax. The encoded messages are prefixed with their length in bytes,
+encoded as an unsigned variable length integer as defined by the [multiformats
+unsigned-varint spec][uvarint-spec].
+
+Messages are encoded via the `ProtoSelect` message type. With the current
+version of _Protocol Select_ detailed in this document, the `version` field of
+the `ProtocolSelect` message is set to `1`. Implementations MUST reject messages
+with a `version` other than the current version. See [Protocol
+Evolution](#Protocol-Evolution) for details. Both the `Offer` and the `Use`
+messages are wrapped with the `ProtocolSelect` message at all time.
+
+```protobuf
+message ProtoSelect {
+    uint32 version = 1;
+
+    message Protocol {
+        oneof protocol {
+            string name = 1;
+        }
+    }
+    repeated Protocol protocols = 2;
+}
+```
+
+<!-- TODO: Document what `name` is. E.g. UTF-8 and same as used in Multistream
+Select. -->
+
+### Protocol Evolution
+
+While we can not foresee all future use-cases of _Protocol Select_, we can
+design _Protocol Select_ in a way to be easy to evolve, and thus be able to
+adapt _Protocol Select_ to support those unknown future use-cases.
+
+#### Non-Breaking Changes
+
+Non-breaking changes to the protocol can be done at the schema level, more
+specifically through the _Protocol Buffer_ framework. Instead of enumerating the
+various update mechanisms, we refer to the _[Updating a Message Type]_ section
+of the _Protocol Buffer_ specification.
+
+As an example for a non-breaking change, say we would like to exchange a made up
+name via the _Protocol Select_ protocol. We can simply extend the `ProtoSelect`
+message type by an `optional string name = 4;` field. Updated implementations
+would be able to extract the name from the payload, old implementations would
+simply ignore the new field.
+
+#### Breaking Changes
+
+When making breaking changes to the _Protocol Select_ protocol,
+implementations need to be able to differentiate the old and the new version on
+the wire. This is done via the `version` field in the `ProtoSelect` message,
+treated as an ordinal monotonically increasing number, with each increase
+identifying a new breaking version of the protocol.
+
+As an example for a made-up breaking change, say we would like the listed
+protocols in the `Offer` message to enumerate the protocols that the local node
```

Comment:
Is this meant to say `ProtoSelect` instead? `Offer` hasn't been mentioned anywhere up till here.

```suggestion
protocols in the `ProtoSelect` message to enumerate the protocols that the local node
```

Reply:
Good catch. Thanks @thomaseizinger.